Development documentation and VSCode configuration

zregvart opened the pull request:
Documents how to run the website preview and how to run/debug the Antora build by using VSCode and `npm link`.

Reference: EC-496

lcarva commented on antora/README.md:
```diff
@@ -1,29 +1,35 @@
-directories as HTML.
+These are the sources for building the https://enterprisecontract.dev website.
+The website uses [Hugo](https://gohugo.io/) and
+[Antora][https://docs.antora.org/] to build different sections of the website.
```
url should be wrapped in `()`, not `[]`.

lcarva commented on antora/README.md:
```diff
@@ -1,29 +1,35 @@
-that the `ec-policies` repository is located at `../ec-policies`. If the local
-repository is on a different path create a symlink pointing from
-`../<repository>` to that path.
+The website makes makes use of
```
Double "makes"

lcarva commented on antora/README.md:
```diff
@@ -1,29 +1,35 @@
-[antora]: https://docs.antora.org/
+Opening the `enterprise-contract.github.io.code-workspace` workspace file within
+VSCode will load the Node projects of the same extensions and the `Antora Build`
+launch configuration in `.vscode/launch.json` allows running the debugger
```
I'm trying this, but my breakpoints on the `index.js` of each extension is ignored. I did run `npm run dev:setup` prior.

zregvart replied:
Do this:
```shell
$ git rev-parse HEAD
6893a5fd05e7c4d193ec04c175e443eca35c0833
$ git clean -dfx
$ cd antora
$ npm run dev:setup

> @enterprise-contract/website@0.0.0 dev:setup
> npm run dev:links && npm link @enterprise-contract/ec-policies-antora-extension @enterprise-contract/reference-antora-extension @enterprise-contract/rego-antora-extension @enterprise-contract/tekton-task-antora-extension


> @enterprise-contract/website@0.0.0 dev:links
> (cd ../../ec-policies/antora/ec-policies-antora-extension && npm link) && (cd ../../ec-cli/reference-antora-extension && npm link) && (cd ../../ec-cli/rego-antora-extension && npm link) && (cd ../../ec-cli/tekton-task-antora-extension && npm link)


up to date, audited 3 packages in 658ms

found 0 vulnerabilities
Reshimming asdf nodejs...

up to date, audited 3 packages in 672ms

found 0 vulnerabilities
Reshimming asdf nodejs...

up to date, audited 3 packages in 663ms

found 0 vulnerabilities
Reshimming asdf nodejs...

up to date, audited 3 packages in 612ms

found 0 vulnerabilities
Reshimming asdf nodejs...

removed 1 package, changed 4 packages, and audited 241 packages in 721ms

34 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities

$ ls -l node_modules/@enterprise-contract # all of these should be symbolic links
total 16
lrwxrwxrwx. 1 zregvart zregvart 59 Mar 26 10:29 ec-policies-antora-extension -> ../../../../ec-policies/antora/ec-policies-antora-extension
lrwxrwxrwx. 1 zregvart zregvart 45 Mar 26 10:29 reference-antora-extension -> ../../../../ec-cli/reference-antora-extension
lrwxrwxrwx. 1 zregvart zregvart 40 Mar 26 10:29 rego-antora-extension -> ../../../../ec-cli/rego-antora-extension
lrwxrwxrwx. 1 zregvart zregvart 47 Mar 26 10:29 tekton-task-antora-extension -> ../../../../ec-cli/tekton-task-antora-extension

$ cd ..
$ code enterprise-contract.github.io.code-workspace
```

Now set a breakpoint in one of the extension's `index.js` files. Go to the Run and Debug tab and run the `Antora Build (enterprise-contract.github.io)` launch configuration.

Do **not** run any other `npm` or `make` commands, they will run `npm ci` which will remove the symlinks created by the `dev:setup` script.

lcarva commented on antora/README.md:
```diff
@@ -1,29 +1,35 @@
-The `hack/local-live.sh` script calls `hack/local-build.sh` for any changes to
-files in the locally cloned repositories. Useful to leave in the background and
-have the generated HTML site continuously updated.
+To make changes locally to the extensions run `npm run dev:setup` script in the
```
I'm probably not doing this right. If I introduce  a typo in `rego-antora-extension/index.js` on the linked ec-cli repo, the build is still successful.

lcarva commented:
It's still confusing to me which sequence of commands I should be running to try out local extension changes.

```
🐚 make
# Run hugo server for website hacking
make hugo-server

# Run hugo server for website hacking with pre-built antora docs
make preview

# Build antora docs once using your locally checked out git repos
make antora-local-build

# Live build antora docs your locally checked out git repos
make antora-local-live

# Build antora docs
make preview-antora

# Build antora docs
make build-antora

# Build antora docs without re-fetching sources
make build-antora-fast

# Build hugo docs
make build-hugo
```

lcarva commented on antora/package.json:
```diff
@@ -22,7 +22,9 @@
     "build": "npm run build:base -- --clean --fetch",
-    "build:fast": "npm run build:base"
+    "build:fast": "npm run build:base",
+    "dev:links": "npm --prefix ../../ec-policies/antora/ec-policies-antora-extension link && npm --prefix ../../ec-cli/reference-antora-extension link && npm --prefix ../../ec-cli/rego-antora-extension link && npm --prefix ../../ec-cli/tekton-task-antora-extension link",
```
`--prefix` here may not be working as expected. If I cd into that directory and run `npm link` then things seem to be working as expected.

zregvart replied:
I noticed that links for some of the dependencies pointed to the global npm install directory, switched to using `cd` instead of `--prefix` and that went away, but introduced another issue, if `npm ci` wasn't invoked in an extension that has a dependency the transitive dependency would not be installed, so I added an additional `npm ci`.

zregvart commented:
Edited the wrong `README.md` I wanted to edit the one in the root of the repository, easier to find it that way.

zregvart committed:
Development documentation and VSCode configuration

Documents how to run the website preview and how to run/debug the Antora
build by using VSCode and `npm link`.

Reference: EC-496

zregvart commented:
> It's still confusing to me which sequence of commands I should be running to try out local extension changes.

None of these, follow the instructions in the README.md they say run `npm run dev:setup` and explicitly call out not running the `preview` target as it will invoke `npm ci` and remove the links setup by `dev:setup`.

lcarva reviewed (approved):
This works for me now. Let's get this merged even though we will eventually move away from using extensions.

zregvart commented:
Might be helpful in the interim...
